ksqldb-common/src/main/java/io/confluent/ksql/function/GenericsUtil.java

```diff
@@ -291,4 +313,27 @@ public static boolean instanceOf(final ParamType schema, final SqlArgument insta
 
     return true;
   }
+
+  public static void reserveGenerics(
```

Comment:
Did you not want to use `reserveGenerics` inside of `UdfIndex` because of the return type + access level (aka the other one is private)? JW because I was confused about `reserveGenerics` being new since it's a function I've seen before. If we leave it here can we come up with another name? IMO all the stuff shared between UDF/Generics is confusing enough w/o duplicate names

Reply:
+1 !

Reply:
I have UdfIndex using this one now, not sure if there's a better way other than having UdfIndex catch the exceptions and return false.

Reply:
Hmm I guess if we really wanted to avoid checking errors in UDFIndex we could have expression type manager call a helper function inside of generics util and have _that_ function call `reserveGenerics` and if `reserveGenerics` returns false it could throw an error... Overall tho I think it'd be nice to not add more small functions that call other functions so I think I lean on the side of leaving it as is

Reply:
Sorry, a few last things here. In this method we call the `resolveGenerics` with two arguments, which does pretty much the exact same thing as this method and throws the same errors. Can we skip the error handling here and leave it in `resolveGenerics`? It's hard for me to think this through logically without running through the code but is there a scenario where we don't catch a generics mismatch in `resolveGenerics` because we don't have the passed in map from `reserveGenerics` and therefore find the exception here? The original `reserveGenerics` did this second check which makes me _think_ it's necessary but I'm not sure.

Reply:
I was originally having issues combining the one in UdfIndex with the one I added since we need to be able to surface the error for invalid generic mappings in the util with lambda functions whereas it's OK to just return true/false in UdfIndex since it's checking candidate functions. I resolved this by having the new function I add return a Pair of boolean and exception to get around this. This still leaves us with a ResolveGeneric and ReserveGeneric in GenericUtils which is definitely confusing, but there's subtle differences to each functions behavior even though it looks like the same logic (`reserve generic` uses an instance of check before resolving, which `resolve generic` doesn't do)

In the end, GenericUtils already had a number of functions that appear to be doing the same logic (with different variations) even before lambdas so I would want to leave an overall refactor of the util out of this pr for now.

Reply:
Never mind, I think I'll be able to cut down on some of the generic util code..

Reply:
Done